shiny/ui/_theme.py

```diff
@@ -0,0 +1,364 @@
+            for key, value in kwargs.items():
+                key.replace("_", "-")
+                if isinstance(value, bool):
+                    value = "true" if value else "false"
+                elif value is None:
+                    value = "null"
+                defaults.append(f"${key}: {value};")
+
+        # Add args to the front of _defaults
+        self._defaults = dedent_array(defaults) + self._defaults
+        self._css = ""
+
+        return self
+
+    def add_mixins(self: T, *args: str) -> T:
+        """
+        Add custom Sass mixins to the theme.
+
+        Sass code added via this method will be placed **after** the mixin declarations
+        from the theme preset, allowing you to override or extend the default mixins.
+
+        Parameters
+        ----------
+        *args
+            Sass code, as a single or multiple strings, containing mixins to add.
+        """
+        self._mixins.extend(dedent_array(args))
+        self._css = ""
+        return self
+
+    def add_rules(self: T, *args: str) -> T:
+        """
+        Add custom Sass rules to the theme.
+
+        Sass code added via this method will be placed **after** the rule declarations
+        from the theme preset, allowing you to override or extend the default rules.
+
+        Parameters
+        ----------
+        *args
+            Sass code, as a single or multiple strings, containing rules to add.
+        """
+        self._rules.extend(dedent_array(args))
+        self._css = ""
+        return self
+
+    def to_sass(self) -> str:
+        """
+        Returns the custom theme as a single Sass string.
+
+        Returns
+        -------
+        :
+            The custom theme as a single Sass string.
+        """
+        path_functions = path_pkg_preset(self._preset, "_01_functions.scss")
+        path_defaults = path_pkg_preset(self._preset, "_02_defaults.scss")
+        path_mixins = path_pkg_preset(self._preset, "_03_mixins.scss")
+        path_rules = path_pkg_preset(self._preset, "_04_rules.scss")
+
+        sass_lines = [
+            f'@import "{path_functions}";',
+            *self._functions,
+            *self._defaults,
+            f'@import "{path_defaults}";',
+            f'@import "{path_mixins}";',
+            *self._mixins,
+            f'@import "{path_rules}";',
+            *self._rules,
+        ]
+
+        return "\n".join(sass_lines)
+
+    def to_css(self) -> str:
+        """
+        Compile the theme to CSS and return the result as a string.
+
+        Returns
+        -------
+        :
+            The compiled CSS for the theme. The value is cached such that previously
+            compiled themes are returned immediately. Adding additional custom Sass code
+            or changing the preset will invalidate the cache.
+        """
+        if self._css:
+            if self._css == "precompiled":
+                return self._read_precompiled_css()
+            return copy.copy(self._css)
```

Comment:
This doesn't seems right. Wouldn't you at least want to check that the Sass input hasn't changed since the last compilation? Also, it doesn't seem we'll be able to avoid false-positive cache hits when `@import` files change on-disk?

(BTW, in R, we crawl the Sass input and get the modified time of import files when building a cache key)

Reply:
Under what conditions are you expecting the import files to change?

Reply:
`.add_rules("@import my-styles.scss")`? Or in our own local development via `Rscript scripts/htmlDependencies.R`?

Reply:
I think there are two important scope-defining characteristics at play here:

1. We intend to render the theme CSS exactly once per app. A `Theme` is provided to `ui.page_*(theme=)` and should only ever be rendered on app start.
2. Most, if not all, relevant file changes happen at the level where they would be detected by `shiny run app.py --reload`. That certainly includes changes from our own local development. We might want to add `*.scss` to the reload watchers, but users could customize `--reload-includes`.

This does bring up an important point, though, that `.add_rules("@import my-styles.scss")` currently doesn't work because of path issues we'd need to work out. I'm very okay with that being follow-up work.